api: Add Viewtype::Vcard and parse_vcard()

iequidoo opened the pull request:
Part of #5422 
Closes #5204

iequidoo commented on src/summary.rs:
```diff
@@ -340,10 +371,6 @@ mod tests {
         msg.set_file("foo.mp3", None);
         assert_summary_texts(&msg, ctx, "🎵 foo.mp3").await; // file name is added for audio
 
-        let mut msg = Message::new(Viewtype::Audio);
```
Removed the duplicated test code, above is the same

iequidoo commented on src/summary.rs:
```diff
@@ -272,6 +287,22 @@ impl Message {
+        let Some(path) = self.get_file(context) else {
+            bail!("No file");
+        };
+        let vcard = tokio::fs::read(path)
```
This should be optimised somehow probably, it's not good that the file is read every time the chatlist is built

Hocuri replied:
If we add a new stock string anyway, we can just always use it. I.e. simply always write "👤 Virtual contact file" (or "👤 Contact", users don't know what a virtual contact file is) into the chatlist. This way, we keep things simpler.

iequidoo replied:
I just tried to do what WhatsApp (and probably others) does, it displays the contact name. Would  be nice if we can do the same. E.g. parse a vCard upon a receipt and save the contact name to the db. Btw, while we don't allow to attach a custom text to vCard messages, we can attach the contact name on the sender side as a temporary solution

iequidoo replied:
Btw, i added a new stock string just in case (and for uniformity with other viewtypes) because if we display the file name as a fallback, the stock string should never be displayed

Hocuri replied:
A new stock string is some work for all the UIs to set it and for translators, so we should add it only with a good reason. I renamed it to "Contact" now which already is a translated string in the UIs (it's just not passed to core), so it's just the UIs that have to set a new stock string and way less work for everyone.

People won't send contacts _that_ often, when they do, it sometimes won't be the last message (that is shown in the chatlist) and when it is, it won't make a big difference whether we write `Contact` or the contact's name. So, let's just call it good enough for now in order to make progress :)

Hocuri commented on src/mimeparser.rs:
```diff
@@ -1936,6 +1937,8 @@ fn get_mime_type(
         mime::AUDIO => Viewtype::Audio,
         mime::VIDEO => Viewtype::Video,
+        mime::VCARD => Viewtype::Vcard,
+        // TODO(vcard): Handle TEXT_VCARD?
```
Not sure what is to do here?

iequidoo replied:
I just didn't get why there are `mime::VCARD` and `mime::TEXT_VCARD`, probably we need to handle the latter because we send vCards as `text/vcard`, see `guess_msgtype_from_suffix()`

Hocuri replied:
Fixed in 2454eaaff, we're matching on `mimetype.type_()` which is `text` in the case of vcards, we need to match on the subtype which is `vcard` in the case of vcards. `mime::TEXT_VCARD` is equal to the string `text/vcard`, which we don't need here.

iequidoo commented on deltachat-jsonrpc/Cargo.toml:
```diff
@@ -15,7 +15,7 @@ required-features = ["webserver"]
 anyhow = "1"
 deltachat = { path = ".." }
-contact-tools = { package = "deltachat-contact-tools", path = "../deltachat-contact-tools" }
+deltachat-contact-tools = { package = "deltachat-contact-tools", path = "../deltachat-contact-tools" }
```
`package = ...` can be removed then

iequidoo reviewed (commented):
Approved!

adbenitez commented on deltachat-contact-tools/src/lib.rs:
```diff
@@ -44,7 +44,8 @@ use regex::Regex;
 pub struct VcardContact {
     /// The email address, vcard property `email`
     pub addr: String,
```
if it is an vcard from the phone's address book, and doesn't have addr, it is just sharing phone number and other data, should it still have vcard view-type or shouldn't it rather be file viewtype since it is not relevant to DC, it is not a DC contact what is being shared :thinking: how is this handled?

Hocuri replied:
Good that you bring that up! I think it should simply be a normal attached file, otherwise we have to tell all UIs that they need to handle this case. This was surprisingly easy to implement: f95273491

adbenitez replied:
great!

iequidoo commented on deltachat-ffi/deltachat.h:
```diff
@@ -5479,6 +5479,11 @@ int64_t         dc_lot_get_timestamp     (const dc_lot_t* lot);
 #define DC_MSG_WEBXDC    80
 
+/**
+ * Message containing a shared contact represented as a vCard (virtual contact file)
```
It can contain multiple contacts, this is already implemented

Hocuri commented on src/mimeparser.rs:
```diff
@@ -1972,6 +1972,19 @@ fn is_attachment_disposition(mail: &mailparse::ParsedMail<'_>) -> bool {
+        return false;
+    };
+    let contacts = deltachat_contact_tools::parse_vcard(body);
+    if let Some(contact) = contacts.first() {
```
It doesn't matter for now since we're concentrating on showing vCards sent by Delta Chat, but note that if we go for "any contact" instead of "first contact", we either need to tell the UI to search through the list for a contact with an email addr, or to filter out contacts without an email, in order to avoid crashes when someone sends a vCard with multiple contacts in it.

BTW, one final other possibility would be to just always return `false` if the vCard doesn't contain exactly one contact. Which I actually like, because then we don't have to worry about very big vCards with lots of contacts, which you brought up in the vcard-contact-sharing group.

adbenitez replied:
> BTW, one final other possibility would be to just always return `false` if the vCard doesn't contain exactly one contact. Which I actually like, because then we don't have to worry about very big vCards with lots of contacts,

+1

iequidoo replied:
Then my change here needs to be just reverted i guess

Hocuri commented on deltachat-jsonrpc/src/api.rs:
```diff
@@ -1426,6 +1427,16 @@ impl CommandApi {
     }
 
+    /// Parses a vCard file located at the given path. Returns contacts in their original order.
+    async fn parse_vcard(&self, path: String) -> Result<Vec<VcardContact>> {
```
> now that deltachat_contact_tools::parse_vcard(vcard) returns a single contact

I don't think that's true?

Of course we could make it return a single contact, but to me the possible churn because we change the API in the future seems worse that the UIs having to always select the the first element.

adbenitez replied:
ah yes sorry it was `Result<Vec<VcardContact>>` the Result what was removed and for some reason I read it as that the Vec was removed from the returning type

iequidoo commented:
So let it be called `make_vcard()`, it's shorter than `create_vcard()` and also "create" usually means "create smth new/empty"

adbenitez reviewed (approved):
I think we can merge this? or is something missing?

adbenitez commented:
> So let it be called `make_vcard()`, it's shorter than `create_vcard()` and also "create" usually means "create smth new/empty"

I think I prefer create_vcard over make_vcard, but it doesn't matter much, if you want to really think deeply about the name then probably something like get_vcard_for_contact() is a better name than the other two names

adbenitez commented:
> I think we can merge this? or is something missing?

(never mind, I was thinking by the name of this PR it was only about adding vcard view type and displaying but now import/export api is also being added in this same PR)

iequidoo commented:
> > I think we can merge this? or is something missing?
> 
> (never mind, I was thinking by the name of this PR it was only about adding vcard view type and displaying but now import/export api is also being added in this same PR)

- Removed the commit adding `make_vcard()`, going to create a separate PR.
- Rebased on top of `main`.
- Squashed commits, please check i did it properly.

Now it can be merged. I think it's better to merge what is already working, at least to avoid further conflicts

Hocuri commented on src/mimeparser.rs:
```diff
@@ -1990,6 +1986,17 @@ fn is_attachment_disposition(mail: &mailparse::ParsedMail<'_>) -> bool {
+        return false;
+    };
+    let contacts = deltachat_contact_tools::parse_vcard(body);
+    if let Some(c) = contacts.first() {
```
```suggestion
    // For now, we only handle vCards containing one single contact:
    if let Some(c) = tools::single_value(contacts) {
```
in order not to have to worry about long vCards with many contacts.

iequidoo replied:
I agree, otherwise a question arises why we ignore the remaining contacts

Hocuri commented on deltachat-contact-tools/src/lib.rs:
```diff
@@ -44,14 +44,25 @@ use regex::Regex;
     pub addr: String,
-    /// The contact's display name, vcard property `fn`
-    pub display_name: String,
+    /// Name authorized by the contact itself, vcard property `fn`. Can be empty, one should use
+    /// `display_name()` to obtain the display name.
+    pub authname: String,
```
Just a note, nothing blocking: It's not actually "authorized by the contact itself", because vCards are sent by a third party (even though AFAIK it's true that the `authname` in `contact.rs` has the same semantics). And the code handling `authname`s in `contact.rs` is a bit far away, while this crate here is only concerned with parsing vCards.

Since i agree that `display_name` is something else in DeltaChat land, I would slightly be in favor of calling this field simply `name`, but feel free to ignore this comment.

iequidoo replied:
I'd avoid using `name`/`display_name` to protect from occasional sharing of locally given names. With `authname` the code in the deltachat core looks just straightforward. But `public_name`/`shared_name` would be also ok. Anyway i agree that the comment should be fixed

Hocuri reviewed (approved):
About the commit messages:
- _Probably_ (not completely sure) the prefix should be `api:`, instead of `feat:` because these two commits not only add a new feature but even a new api
- "[feat(jsonrpc): Add a function parsing a vCard file at the given path](https://github.com/deltachat/deltachat-core-rust/pull/5536/commits/05b775cd447024c1eb63d1ff7690389a0611bab1)": This should mention the name of the added function, i.e. something like `api(jsonrpc): Add parse_vcard() that can parse the file attached to a Viewtype::Vcard` or `api(jsonrpc): Add parse_vcard()` or so

Christianpaciel commented:
Dale fotto.  Seguro.  Mo. Es tuyo

Christianpaciel commented:
Forro   segui.   Con tus chat  Forro
